internal/anchore/client.go

```diff
@@ -58,26 +60,56 @@ func (c *Client) newRequestContext(parentContext context.Context) context.Contex
 	)
 }
 
-func hasScheme(url string) bool {
-	parts := strings.Split(url, "://")
+var ErrInvalidBaseURLInput = errors.New("invalid base URL input")
 
-	return len(parts) > 1
+func prepareBaseURLForClient(baseURL string) (string, error) {
+	if err := checkBaseURLInput(baseURL); err != nil {
+		return "", err
+	}
+
+	scheme, urlWithoutScheme := splitSchemeFromURL(baseURL)
+
+	if scheme == "" {
+		scheme = "http"
+	}
+
+	urlWithoutScheme = path.Clean(urlWithoutScheme)
+
+	const requiredSuffix = "v1"
+	if path.Base(urlWithoutScheme) != requiredSuffix {
+		urlWithoutScheme = path.Join(urlWithoutScheme, requiredSuffix)
+	}
+
+	preparedBaseURL := scheme + "://" + urlWithoutScheme
+	return preparedBaseURL, nil
 }
 
-func ensureURLHasScheme(url string) string {
-	const defaultScheme = "http"
+func checkBaseURLInput(url string) error {
+	if url == "" {
+		return ErrInvalidBaseURLInput
+	}
 
-	if !hasScheme(url) {
-		return fmt.Sprintf("%s://%s", defaultScheme, url)
+	firstCharacter := rune(url[0])
+	if !(unicode.IsLetter(firstCharacter)) {
+		return ErrInvalidBaseURLInput
 	}
 
-	return url
+	return nil
 }
 
-func ensureURLHasSuffix(url, suffix string) string {
-	if !strings.HasSuffix(url, suffix) {
-		return url + suffix
+func splitSchemeFromURL(url string) (scheme, urlWithoutScheme string) {
+	if hasScheme(url) {
+		urlParts := strings.SplitN(url, "://", 2)
+		scheme = urlParts[0]
+		urlWithoutScheme = urlParts[1]
+		return
 	}
 
-	return url
+	return "", url
```

Comment:
since the consumer of this function will always want a scheme and a url, shouldn't the fallback to `"http"` be used here instead of being done in the caller? (not a hard comment, just curious about the alternative)
```suggestion
	return "http", url
```

Reply:
@alfredodeza It's funny you say that, I was thinking about doing that initially. Then I reflected and thought the process was simpler when each step was broken out — like "Okay, what's the scheme? Okay, the scheme is nothing, so here's what we'll use". This way, the function that splits the scheme off is transparent, and the consumer can decide what the default is. It's important to have a default, for sure, but this approach segregates the responsibilities of using a default (like HTTP) vs. breaking the scheme off from the rest of the URL (which is its own important operation)